linkerd/app/core/src/addr_match.rs

```diff
@@ -0,0 +1,59 @@
+use ipnet::{Contains, IpNet};
+use linkerd2_addr::Addr;
+use linkerd2_dns::{Name, Suffix};
+use std::{net::IpAddr, sync::Arc};
+
+#[derive(Clone, Debug)]
+pub struct AddrMatch(NameMatch, IpMatch);
+
+#[derive(Clone, Debug)]
+pub struct NameMatch(Arc<Vec<Suffix>>);
+
+#[derive(Clone, Debug)]
+pub struct IpMatch(Arc<Vec<IpNet>>);
+
+// === impl NameMatch ===
+
+impl AddrMatch {
+    pub fn new(
+        suffixes: impl IntoIterator<Item = Suffix>,
+        nets: impl IntoIterator<Item = IpNet>,
+    ) -> Self {
+        Self(NameMatch::new(suffixes), IpMatch::new(nets))
+    }
+
+    pub fn matches(&self, addr: &Addr) -> bool {
+        match addr {
+            Addr::Name(ref name) => self.0.matches(name.name()),
+            Addr::Socket(sa) => self.1.matches(sa.ip()),
```

Comment:
nit/TIOLI: this might read nicer if we named the two fields of `AddrMatch`, but that's not a huge deal...